jdaviz/configs/imviz/plugins/viewers.py

```diff
@@ -162,15 +168,47 @@ def blink_once(self):
                 next_layer = valid[(valid.index(visible[-1]) + 1) % n_layers]
                 self.state.layers[next_layer].visible = True
 
-                # TODO: We can display the active data label in GUI here.
-                # For now, we will print to debug Output widget.
-                data_label = self.state.layers[next_layer].layer.label
-                with self.session.application.output:
-                    print(f'Visible layer changed to: {data_label}')
-
                 for ilayer in (set(valid) - set([next_layer])):
                     self.state.layers[ilayer].visible = False
 
+                # We can display the active data label in Compass plugin.
+                self.set_compass(self.state.layers[next_layer].layer)
+
+    def on_limits_change(self, *args):
+        try:
+            i = get_top_layer_index(self)
+        except IndexError:
+            return
+        self.set_compass(self.state.layers[i].layer)
+
+    def set_compass(self, image):
+        """Update the Compass plugin with info from the given image Data object."""
+        if self.compass is None:
+            try:
+                self.compass = self.session.parent_app.get_tray_item_from_name('imviz-compass')
```

Comment:
Following https://github.com/spacetelescope/jdaviz/pull/983/files#r792096116 , I am unable to access the "parent_app" here even though I attached it to the viewer object on creation. What am I doing wrong? Here is the traceback:

```
.../echo/core.py", line 125, in notify
    cback(new)
  File ".../jdaviz/jdaviz/configs/imviz/plugins/viewers.py", line 182, in on_limits_change
    self.set_compass(self.state.layers[i].layer)
  File ".../jdaviz/jdaviz/configs/imviz/plugins/viewers.py", line 188, in set_compass
    self.compass = self.session.parent_app.get_tray_item_from_name('imviz-compass')
AttributeError: 'Session' object has no attribute 'parent_app'
ERROR:tornado.application:Exception in callback functools.partial(<function debounced.<locals>.wrapped.<locals>.execute.<locals>.debounced_execute at 0x7f16f45f4d30>)
Traceback (most recent call last):
  File ".../tornado/ioloop.py", line 741, in _run_callback
    ret = callback()
  File ".../glue-jupyter/glue_jupyter/utils.py", line 134, in debounced_execute
    f(*args, **kwargs)
  File ".../glue-jupyter/glue_jupyter/bqplot/common/viewer.py", line 175, in update_glue_scales
    self.state.update_from_dict(state)
  File ".../glue/glue/core/state_objects.py", line 72, in update_from_dict
    setattr(self, name, properties[name])
  File ".../echo/core.py", line 538, in __exit__
    p.notify(*args)
  File ".../glue/glue/utils/matplotlib.py", line 170, in wrapper
    result = func(*args, **kwargs)
  File ".../glue/glue/viewers/matplotlib/state.py", line 24, in notify
    super(DeferredDrawCallbackProperty, self).notify(*args, **kwargs)
  File ".../echo/core.py", line 125, in notify
    cback(new)
  File ".../glue/glue/viewers/matplotlib/state.py", line 162, in _adjust_limits_aspect_x
    self._adjust_limits_aspect(aspect_adjustable='y')
  File ".../glue/glue/utils/decorators.py", line 32, in wrapper
    return meth(self, *args, **kwargs)
  File ".../glue/glue/viewers/matplotlib/state.py", line 235, in _adjust_limits_aspect
    self.y_max = y_max
  File ".../echo/core.py", line 538, in __exit__
    p.notify(*args)
  File ".../glue/glue/utils/matplotlib.py", line 165, in wrapper
    return func(*args, **kwargs)
  File ".../glue/glue/viewers/matplotlib/state.py", line 24, in notify
    super(DeferredDrawCallbackProperty, self).notify(*args, **kwargs)
  File ".../echo/core.py", line 125, in notify
    cback(new)
  File ".../jdaviz/jdaviz/configs/imviz/plugins/viewers.py", line 182, in on_limits_change
    self.set_compass(self.state.layers[i].layer)
  File ".../jdaviz/jdaviz/configs/imviz/plugins/viewers.py", line 188, in set_compass
    self.compass = self.session.parent_app.get_tray_item_from_name('imviz-compass')
AttributeError: 'Session' object has no attribute 'parent_app'
ERROR:tornado.application:Exception in callback functools.partial(<function debounced.<locals>.wrapped.<locals>.execute.<locals>.debounced_execute at 0x7f16f45f4c10>)
Traceback (most recent call last):
  File ".../tornado/ioloop.py", line 741, in _run_callback
    ret = callback()
  File ".../glue-jupyter/glue_jupyter/utils.py", line 134, in debounced_execute
    f(*args, **kwargs)
  File ".../glue-jupyter/glue_jupyter/bqplot/common/viewer.py", line 175, in update_glue_scales
    self.state.update_from_dict(state)
  File ".../glue/glue/core/state_objects.py", line 72, in update_from_dict
    setattr(self, name, properties[name])
  File ".../echo/core.py", line 538, in __exit__
    p.notify(*args)
  File ".../glue/glue/utils/matplotlib.py", line 170, in wrapper
    result = func(*args, **kwargs)
  FiERROR:tornado.application:Exception in callback functools.partial(<function debounced.<locals>.wrapped.<locals>.execute.<locals>.debounced_execute at 0x7f16f45f4c10>)
Traceback (most recent call last):
  File ".../tornado/ioloop.py", line 741, in _run_callback
    ret = callback()
  File ".../glue-jupyter/glue_jupyter/utils.py", line 134, in debounced_execute
    f(*args, **kwargs)
  File ".../glue-jupyter/glue_jupyter/bqplot/common/viewer.py", line 175, in update_glue_scales
    self.state.update_from_dict(state)
  File ".../glue/glue/core/state_objects.py", line 72, in update_from_dict
    setattr(self, name, properties[name])
  File ".../echo/core.py", line 538, in __exit__
    p.notify(*args)
  File ".../glue/glue/utils/matplotlib.py", line 170, in wrapper
    result = func(*args, **kwargs)
  File ".../glue/glue/viewers/matplotlib/state.py", line 24, in notify
    super(DeferredDrawCallbackProperty, self).notify(*args, **kwargs)
  File ".../echo/core.py", line 125, in notify
    cback(new)
  File ".../glue/glue/viewers/matplotlib/state.py", line 165, in _adjust_limits_aspect_y
    self._adjust_limits_aspect(aspect_adjustable='x')
  File ".../glue/glue/utils/decorators.py", line 32, in wrapper
    return meth(self, *args, **kwargs)
  File ".../glue/glue/viewers/matplotlib/state.py", line 235, in _adjust_limits_aspect
    self.y_max = y_max
  File ".../echo/core.py", line 538, in __exit__
    p.notify(*args)
  File ".../glue/glue/utils/matplotlib.py", line 165, in wrapper
    return func(*args, **kwargs)
  File ".../glue/glue/viewers/matplotlib/state.py", line 24, in notify
    super(DeferredDrawCallbackProperty, self).notify(*args, **kwargs)
  File ".../echo/core.py", line 125, in notify
    cback(new)
  File ".../jdaviz/jdaviz/configs/imviz/plugins/viewers.py", line 182, in on_limits_change
    self.set_compass(self.state.layers[i].layer)
  File ".../jdaviz/jdaviz/configs/imviz/plugins/viewers.py", line 188, in set_compass
    self.compass = self.session.parent_app.get_tray_item_from_name('imviz-compass')
AttributeError: 'Session' object has no attribute 'parent_app'
le ".../glue/glue/viewers/matplotlib/state.py", line 24, in notify
    super(DeferredDrawCallbackProperty, self).notify(*args, **kwargs)
  File ".../echo/core.py", line 125, in notify
    cback(new)
  File ".../glue/glue/viewers/matplotlib/state.py", line 165, in _adjust_limits_aspect_y
    self._adjust_limits_aspect(aspect_adjustable='x')
  File ".../glue/glue/utils/decorators.py", line 32, in wrapper
    return meth(self, *args, **kwargs)
  File ".../glue/glue/viewers/matplotlib/state.py", line 235, in _adjust_limits_aspect
    self.y_max = y_max
  File ".../echo/core.py", line 538, in __exit__
    p.notify(*args)
  File ".../glue/glue/utils/matplotlib.py", line 165, in wrapper
    return func(*args, **kwargs)
  File ".../glue/glue/viewers/matplotlib/state.py", line 24, in notify
    super(DeferredDrawCallbackProperty, self).notify(*args, **kwargs)
  File ".../echo/core.py", line 125, in notify
    cback(new)
  File ".../jdaviz/jdaviz/configs/imviz/plugins/viewers.py", line 182, in on_limits_change
    self.set_compass(self.state.layers[i].layer)
  File ".../jdaviz/jdaviz/configs/imviz/plugins/viewers.py", line 188, in set_compass
    self.compass = self.session.parent_app.get_tray_item_from_name('imviz-compass')
AttributeError: 'Session' object has no attribute 'parent_app'
le ".../glue/glue/viewers/matplotlib/state.py", line 24, in notify
    super(DeferredDrawCallbackProperty, self).notify(*args, **kwargs)
  File ".../echo/core.py", line 125, in notify
    cback(new)
  File ".../glue/glue/viewers/matplotlib/state.py", line 162, in _adjust_limits_aspect_x
    self._adjust_limits_aspect(aspect_adjustable='y')
  File ".../glue/glue/utils/decorators.py", line 32, in wrapper
    return meth(self, *args, **kwargs)
  File ".../glue/glue/viewers/matplotlib/state.py", line 235, in _adjust_limits_aspect
    self.y_max = y_max
  File ".../echo/core.py", line 538, in __exit__
    p.notify(*args)
  File ".../glue/glue/utils/matplotlib.py", line 165, in wrapper
    return func(*args, **kwargs)
  File ".../glue/glue/viewers/matplotlib/state.py", line 24, in notify
    super(DeferredDrawCallbackProperty, self).notify(*args, **kwargs)
  File ".../echo/core.py", line 125, in notify
    cback(new)
  File ".../jdaviz/jdaviz/configs/imviz/plugins/viewers.py", line 182, in on_limits_change
    self.set_compass(self.state.layers[i].layer)
  File ".../jdaviz/jdaviz/configs/imviz/plugins/viewers.py", line 188, in set_compass
    self.compass = self.session.parent_app.get_tray_item_from_name('imviz-compass')
AttributeError: 'Session' object has no attribute 'parent_app'
ERROR:tornado.application:Exception in callback functools.partial(<function debounced.<locals>.wrapped.<locals>.execute.<locals>.debounced_execute at 0x7f16f4666b80>)
Traceback (most recent call last):
  File ".../tornado/ioloop.py", line 741, in _run_callback
    ret = callback()
  File ".../glue-jupyter/glue_jupyter/utils.py", line 134, in debounced_execute
    f(*args, **kwargs)
  File ".../glue-jupyter/glue_jupyter/bqplot/common/viewer.py", line 175, in update_glue_scales
    self.state.update_from_dict(state)
  File ".../glue/glue/core/state_objects.py", line 72, in update_from_dict
    setattr(self, name, properties[name])
  File ".../echo/core.py", line 538, in __exit__
    p.notify(*args)
  File ".../glue/glue/utils/matplotlib.py", line 170, in wrapper
    result = func(*args, **kwargs)
  File ".../glue/glue/viewers/matplotlib/state.py", line 24, in notify
    super(DeferredDrawCallbackProperty, self).notify(*args, **kwargs)
  File ".../echo/core.py", line 125, in notify
    cback(new)
  File ".../glue/glue/viewers/matplotlib/state.py", line 165, in _adjust_limits_aspect_y
    self._adjust_limits_aspect(aspect_adjustable='x')
  File ".../glue/glue/utils/decorators.py", line 32, in wrapper
    return meth(self, *args, **kwargs)
  File ".../glue/glue/viewers/matplotlib/state.py", line 235, in _adjust_limits_aspect
    self.y_max = y_max
  File ".../echo/core.py", line 538, in __exit__
    p.notify(*args)
  File ".../glue/glue/utils/matplotlib.py", line 165, in wrapper
    return func(*args, **kwargs)
  File ".../glue/glue/viewers/matplotlib/state.py", line 24, in notify
    super(DeferredDrawCallbackProperty, self).notify(*args, **kwargs)
  File ".../echo/core.py", line 125, in notify
    cback(new)
  File ".../jdaviz/jdaviz/configs/imviz/plugins/viewers.py", line 182, in on_limits_change
    self.set_compass(self.state.layers[i].layer)
  File ".../jdaviz/jdaviz/configs/imviz/plugins/viewers.py", line 188, in set_compass
    self.compass = self.session.parent_app.get_tray_item_from_name('imviz-compass')
AttributeError: 'Session' object has no attribute 'parent_app'
ERROR:tornado.application:Exception in callback functools.partial(<function debounced.<locals>.wrapped.<locals>.execute.<locals>.debounced_execute at 0x7f16f4666b80>)
Traceback (most recent call last):
  File ".../tornado/ioloop.py", line 741, in _run_callback
    ret = callback()
  File ".../glue-jupyter/glue_jupyter/utils.py", line 134, in debounced_execute
    f(*args, **kwargs)
  File ".../glue-jupyter/glue_jupyter/bqplot/common/viewer.py", line 175, in update_glue_scales
    self.state.update_from_dict(state)
  File ".../glue/glue/core/state_objects.py", line 72, in update_from_dict
    setattr(self, name, properties[name])
  File ".../echo/core.py", line 538, in __exit__
    p.notify(*args)
  File ".../glue/glue/utils/matplotlib.py", line 170, in wrapper
    result = func(*args, **kwargs)
  File ".../glue/glue/viewers/matplotlib/state.py", line 24, in notify
    super(DeferredDrawCallbackProperty, self).notify(*args, **kwargs)
  File ".../echo/core.py", line 125, in notify
    cback(new)
  File ".../glue/glue/viewers/matplotlib/state.py", line 165, in _adjust_limits_aspect_y
    self._adjust_limits_aspect(aspect_adjustable='x')
  File ".../glue/glue/utils/decorators.py", line 32, in wrapper
    return meth(self, *args, **kwargs)
  File ".../glue/glue/viewers/matplotlib/state.py", line 235, in _adjust_limits_aspect
    self.y_max = y_max
  File ".../echo/core.py", line 538, in __exit__
    p.notify(*args)
  File ".../glue/glue/utils/matplotlib.py", line 165, in wrapper
    return func(*args, **kwargs)
  File ".../glue/glue/viewers/matplotlib/state.py", line 24, in notify
    super(DeferredDrawCallbackProperty, self).notify(*args, **kwargs)
  File ".../echo/core.py", line 125, in notify
    cback(new)
  File ".../jdaviz/jdaviz/configs/imviz/plugins/viewers.py", line 182, in on_limits_change
    self.set_compass(self.state.layers[i].layer)
  File ".../jdaviz/jdaviz/configs/imviz/plugins/viewers.py", line 188, in set_compass
    self.compass = self.session.parent_app.get_tray_item_from_name('imviz-compass')
AttributeError: 'Session' object has no attribute 'parent_app'
ERROR:tornado.application:Exception in callback functools.partial(<function debounced.<locals>.wrapped.<locals>.execute.<locals>.debounced_execute at 0x7f16f45f4dc0>)
Traceback (most recent call last):
  File ".../tornado/ioloop.py", line 741, in _run_callback
    ret = callback()
  File ".../glue-jupyter/glue_jupyter/utils.py", line 134, in debounced_execute
    f(*args, **kwargs)
  File ".../glue-jupyter/glue_jupyter/bqplot/common/viewer.py", line 175, in update_glue_scales
    self.state.update_from_dict(state)
  File ".../glue/glue/core/state_objects.py", line 72, in update_from_dict
    setattr(self, name, properties[name])
  File ".../echo/core.py", line 538, in __exit__
    p.notify(*args)
  File ".../glue/glue/utils/matplotlib.py", line 170, in wrapper
    result = func(*args, **kwargs)
  File ".../glue/glue/viewers/matplotlib/state.py", line 24, in notify
    super(DeferredDrawCallbackProperty, self).notify(*args, **kwargs)
  File ".../echo/core.py", line 125, in notify
    cback(new)
  File ".../glue/glue/viewers/matplotlib/state.py", line 162, in _adjust_limits_aspect_x
    self._adjust_limits_aspect(aspect_adjustable='y')
  File ".../glue/glue/utils/decorators.py", line 32, in wrapper
    return meth(self, *args, **kwargs)
  File ".../glue/glue/viewers/matplotlib/state.py", line 235, in _adjust_limits_aspect
    self.y_max = y_max
  File ".../echo/core.py", line 538, in __exit__
    p.notify(*args)
  File ".../glue/glue/utils/matplotlib.py", line 165, in wrapper
    return func(*args, **kwargs)
  File ".../glue/glue/viewers/matplotlib/state.py", line 24, in notify
    super(DeferredDrawCallbackProperty, self).notify(*args, **kwargs)
  File ".../echo/core.py", line 125, in notify
    cback(new)
  File ".../jdaviz/jdaviz/configs/imviz/plugins/viewers.py", line 182, in on_limits_change
    self.set_compass(self.state.layers[i].layer)
  File ".../jdaviz/jdaviz/configs/imviz/plugins/viewers.py", line 188, in set_compass
    self.compass = self.session.parent_app.get_tray_item_from_name('imviz-compass')
AttributeError: 'Session' object has no attribute 'parent_app'
ERROR:tornado.application:Exception in callback functools.partial(<function debounced.<locals>.wrapped.<locals>.execute.<locals>.debounced_execute at 0x7f16f45f4dc0>)
Traceback (most recent call last):
  File ".../tornado/ioloop.py", line 741, in _run_callback
    ret = callback()
  File ".../glue-jupyter/glue_jupyter/utils.py", line 134, in debounced_execute
    f(*args, **kwargs)
  File ".../glue-jupyter/glue_jupyter/bqplot/common/viewer.py", line 175, in update_glue_scales
    self.state.update_from_dict(state)
  File ".../glue/glue/core/state_objects.py", line 72, in update_from_dict
    setattr(self, name, properties[name])
  File ".../echo/core.py", line 538, in __exit__
    p.notify(*args)
  File ".../glue/glue/utils/matplotlib.py", line 170, in wrapper
    result = func(*args, **kwargs)
  File ".../glue/glue/viewers/matplotlib/state.py", line 24, in notify
    super(DeferredDrawCallbackProperty, self).notify(*args, **kwargs)
  File ".../echo/core.py", line 125, in notify
    cback(new)
  File ".../glue/glue/viewers/matplotlib/state.py", line 162, in _adjust_limits_aspect_x
    self._adjust_limits_aspect(aspect_adjustable='y')
  File ".../glue/glue/utils/decorators.py", line 32, in wrapper
    return meth(self, *args, **kwargs)
  File ".../glue/glue/viewers/matplotlib/state.py", line 235, in _adjust_limits_aspect
    self.y_max = y_max
  File ".../echo/core.py", line 538, in __exit__
    p.notify(*args)
  File ".../glue/glue/utils/matplotlib.py", line 165, in wrapper
    return func(*args, **kwargs)
  File ".../glue/glue/viewers/matplotlib/state.py", line 24, in notify
    super(DeferredDrawCallbackProperty, self).notify(*args, **kwargs)
  File ".../echo/core.py", line 125, in notify
    cback(new)
  File ".../jdaviz/jdaviz/configs/imviz/plugins/viewers.py", line 182, in on_limits_change
    self.set_compass(self.state.layers[i].layer)
  File ".../jdaviz/jdaviz/configs/imviz/plugins/viewers.py", line 188, in set_compass
    self.compass = self.session.parent_app.get_tray_item_from_name('imviz-compass')
AttributeError: 'Session' object has no attribute 'parent_app'
ERROR:tornado.application:Exception in callback functools.partial(<function debounced.<locals>.wrapped.<locals>.execute.<locals>.debounced_execute at 0x7f16f4666b80>)
Traceback (most recent call last):
  File ".../tornado/ioloop.py", line 741, in _run_callback
    ret = callback()
  File ".../glue-jupyter/glue_jupyter/utils.py", line 134, in debounced_execute
    f(*args, **kwargs)
  File ".../glue-jupyter/glue_jupyter/bqplot/common/viewer.py", line 175, in update_glue_scales
    self.state.update_from_dict(state)
  File ".../glue/glue/core/state_objects.py", line 72, in update_from_dict
    setattr(self, name, properties[name])
  File ".../echo/core.py", line 538, in __exit__
    p.notify(*args)
  File ".../glue/glue/utils/matplotlib.py", line 170, in wrapper
    result = func(*args, **kwargs)
  File ".../glue/glue/viewers/matplotlib/state.py", line 24, in notify
    super(DeferredDrawCallbackProperty, self).notify(*args, **kwargs)
  File ".../echo/core.py", line 125, in notify
    cback(new)
  File ".../glue/glue/viewers/matplotlib/state.py", line 165, in _adjust_limits_aspect_y
    self._adjust_limits_aspect(aspect_adjustable='x')
  File ".../glue/glue/utils/decorators.py", line 32, in wrapper
    return meth(self, *args, **kwargs)
  File ".../glue/glue/viewers/matplotlib/state.py", line 235, in _adjust_limits_aspect
    self.y_max = y_max
  File ".../echo/core.py", line 538, in __exit__
    p.notify(*args)
  File ".../glue/glue/utils/matplotlib.py", line 165, in wrapper
    return func(*args, **kwargs)
  File ".../glue/glue/viewers/matplotlib/state.py", line 24, in notify
    super(DeferredDrawCallbackProperty, self).notify(*args, **kwargs)
  File ".../echo/core.py", line 125, in notify
    cback(new)
  File ".../jdaviz/jdaviz/configs/imviz/plugins/viewers.py", line 182, in on_limits_change
    self.set_compass(self.state.layers[i].layer)
  File ".../jdaviz/jdaviz/configs/imviz/plugins/viewers.py", line 188, in set_compass
    self.compass = self.session.parent_app.get_tray_item_from_name('imviz-compass')
AttributeError: 'Session' object has no attribute 'parent_app'
ERROR:tornado.application:Exception in callback functools.partial(<function debounced.<locals>.wrapped.<locals>.execute.<locals>.debounced_execute at 0x7f16f4666b80>)
Traceback (most recent call last):
  File ".../tornado/ioloop.py", line 741, in _run_callback
    ret = callback()
  File ".../glue-jupyter/glue_jupyter/utils.py", line 134, in debounced_execute
    f(*args, **kwargs)
  File ".../glue-jupyter/glue_jupyter/bqplot/common/viewer.py", line 175, in update_glue_scales
    self.state.update_from_dict(state)
  File ".../glue/glue/core/state_objects.py", line 72, in update_from_dict
    setattr(self, name, properties[name])
  File ".../echo/core.py", line 538, in __exit__
    p.notify(*args)
  File ".../glue/glue/utils/matplotlib.py", line 170, in wrapper
    result = func(*args, **kwargs)
  File ".../glue/glue/viewers/matplotlib/state.py", line 24, in notify
    super(DeferredDrawCallbackProperty, self).notify(*args, **kwargs)
  File ".../echo/core.py", line 125, in notify
    cback(new)
  File ".../glue/glue/viewers/matplotlib/state.py", line 165, in _adjust_limits_aspect_y
    self._adjust_limits_aspect(aspect_adjustable='x')
  File ".../glue/glue/utils/decorators.py", line 32, in wrapper
    return meth(self, *args, **kwargs)
  File ".../glue/glue/viewers/matplotlib/state.py", line 235, in _adjust_limits_aspect
    self.y_max = y_max
  File ".../echo/core.py", line 538, in __exit__
    p.notify(*args)
  File ".../glue/glue/utils/matplotlib.py", line 165, in wrapper
    return func(*args, **kwargs)
  File ".../glue/glue/viewers/matplotlib/state.py", line 24, in notify
    super(DeferredDrawCallbackProperty, self).notify(*args, **kwargs)
  File ".../echo/core.py", line 125, in notify
    cback(new)
  File ".../jdaviz/jdaviz/configs/imviz/plugins/viewers.py", line 182, in on_limits_change
    self.set_compass(self.state.layers[i].layer)
  File ".../jdaviz/jdaviz/configs/imviz/plugins/viewers.py", line 188, in set_compass
    self.compass = self.session.parent_app.get_tray_item_from_name('imviz-compass')
AttributeError: 'Session' object has no attribute 'parent_app'
ERROR:tornado.application:Exception in callback functools.partial(<function debounced.<locals>.wrapped.<locals>.execute.<locals>.debounced_execute at 0x7f16f45f4ee0>)
Traceback (most recent call last):
  File ".../tornado/ioloop.py", line 741, in _run_callback
    ret = callback()
  File ".../glue-jupyter/glue_jupyter/utils.py", line 134, in debounced_execute
    f(*args, **kwargs)
  File ".../glue-jupyter/glue_jupyter/bqplot/common/viewer.py", line 175, in update_glue_scales
    self.state.update_from_dict(state)
  File ".../glue/glue/core/state_objects.py", line 72, in update_from_dict
    setattr(self, name, properties[name])
  File ".../echo/core.py", line 538, in __exit__
    p.notify(*args)
  File ".../glue/glue/utils/matplotlib.py", line 170, in wrapper
    result = func(*args, **kwargs)
  File ".../glue/glue/viewers/matplotlib/state.py", line 24, in notify
    super(DeferredDrawCallbackProperty, self).notify(*args, **kwargs)
  File ".../echo/core.py", line 125, in notify
    cback(new)
  File ".../glue/glue/viewers/matplotlib/state.py", line 162, in _adjust_limits_aspect_x
    self._adjust_limits_aspect(aspect_adjustable='y')
  File ".../glue/glue/utils/decorators.py", line 32, in wrapper
    return meth(self, *args, **kwargs)
  File ".../glue/glue/viewers/matplotlib/state.py", line 235, in _adjust_limits_aspect
    self.y_max = y_max
  File ".../echo/core.py", line 538, in __exit__
    p.notify(*args)
  File ".../glue/glue/utils/matplotlib.py", line 165, in wrapper
    return func(*args, **kwargs)
  File ".../glue/glue/viewers/matplotlib/state.py", line 24, in notify
    super(DeferredDrawCallbackProperty, self).notify(*args, **kwargs)
  File ".../echo/core.py", line 125, in notify
    cback(new)
  File ".../jdaviz/jdaviz/configs/imviz/plugins/viewers.py", line 182, in on_limits_change
    self.set_compass(self.state.layers[i].layer)
  File ".../jdaviz/jdaviz/configs/imviz/plugins/viewers.py", line 188, in set_compass
    self.compass = self.session.parent_app.get_tray_item_from_name('imviz-compass')
AttributeError: 'Session' object has no attribute 'parent_app'
ERROR:tornado.application:Exception in callback functools.partial(<function debounced.<locals>.wrapped.<locals>.execute.<locals>.debounced_execute at 0x7f16f45f4ee0>)
Traceback (most recent call last):
  File ".../tornado/ioloop.py", line 741, in _run_callback
    ret = callback()
  File ".../glue-jupyter/glue_jupyter/utils.py", line 134, in debounced_execute
    f(*args, **kwargs)
  File ".../glue-jupyter/glue_jupyter/bqplot/common/viewer.py", line 175, in update_glue_scales
    self.state.update_from_dict(state)
  File ".../glue/glue/core/state_objects.py", line 72, in update_from_dict
    setattr(self, name, properties[name])
  File ".../echo/core.py", line 538, in __exit__
    p.notify(*args)
  File ".../glue/glue/utils/matplotlib.py", line 170, in wrapper
    result = func(*args, **kwargs)
  File ".../glue/glue/viewers/matplotlib/state.py", line 24, in notify
    super(DeferredDrawCallbackProperty, self).notify(*args, **kwargs)
  File ".../echo/core.py", line 125, in notify
    cback(new)
  File ".../glue/glue/viewers/matplotlib/state.py", line 162, in _adjust_limits_aspect_x
    self._adjust_limits_aspect(aspect_adjustable='y')
  File ".../glue/glue/utils/decorators.py", line 32, in wrapper
    return meth(self, *args, **kwargs)
  File ".../glue/glue/viewers/matplotlib/state.py", line 235, in _adjust_limits_aspect
    self.y_max = y_max
  File ".../echo/core.py", line 538, in __exit__
    p.notify(*args)
  File ".../glue/glue/utils/matplotlib.py", line 165, in wrapper
    return func(*args, **kwargs)
  File ".../glue/glue/viewers/matplotlib/state.py", line 24, in notify
    super(DeferredDrawCallbackProperty, self).notify(*args, **kwargs)
  File ".../echo/core.py", line 125, in notify
    cback(new)
  File ".../jdaviz/jdaviz/configs/imviz/plugins/viewers.py", line 182, in on_limits_change
    self.set_compass(self.state.layers[i].layer)
  File ".../jdaviz/jdaviz/configs/imviz/plugins/viewers.py", line 188, in set_compass
    self.compass = self.session.parent_app.get_tray_item_from_name('imviz-compass')
AttributeError: 'Session' object has no attribute 'parent_app'
ERROR:tornado.application:Exception in callback functools.partial(<function debounced.<locals>.wrapped.<locals>.execute.<locals>.debounced_execute at 0x7f16f4666b80>)
Traceback (most recent call last):
  File ".../tornado/ioloop.py", line 741, in _run_callback
    ret = callback()
  File ".../glue-jupyter/glue_jupyter/utils.py", line 134, in debounced_execute
    f(*args, **kwargs)
  File ".../glue-jupyter/glue_jupyter/bqplot/common/viewer.py", line 175, in update_glue_scales
    self.state.update_from_dict(state)
  File ".../glue/glue/core/state_objects.py", line 72, in update_from_dict
    setattr(self, name, properties[name])
  File ".../echo/core.py", line 538, in __exit__
    p.notify(*args)
  File ".../glue/glue/utils/matplotlib.py", line 170, in wrapper
    result = func(*args, **kwargs)
  File ".../glue/glue/viewers/matplotlib/state.py", line 24, in notify
    super(DeferredDrawCallbackProperty, self).notify(*args, **kwargs)
  File ".../echo/core.py", line 125, in notify
    cback(new)
  File ".../jdaviz/jdaviz/configs/imviz/plugins/viewers.py", line 182, in on_limits_change
    self.set_compass(self.state.layers[i].layer)
  File ".../jdaviz/jdaviz/configs/imviz/plugins/viewers.py", line 188, in set_compass
    self.compass = self.session.parent_app.get_tray_item_from_name('imviz-compass')
AttributeError: 'Session' object has no attribute 'parent_app'
ERROR:tornado.application:Exception in callback functools.partial(<function debounced.<locals>.wrapped.<locals>.execute.<locals>.debounced_execute at 0x7f16f4666b80>)
Traceback (most recent call last):
  File ".../tornado/ioloop.py", line 741, in _run_callback
    ret = callback()
  File ".../glue-jupyter/glue_jupyter/utils.py", line 134, in debounced_execute
    f(*args, **kwargs)
  File ".../glue-jupyter/glue_jupyter/bqplot/common/viewer.py", line 175, in update_glue_scales
    self.state.update_from_dict(state)
  File ".../glue/glue/core/state_objects.py", line 72, in update_from_dict
    setattr(self, name, properties[name])
  File ".../echo/core.py", line 538, in __exit__
    p.notify(*args)
  File ".../glue/glue/utils/matplotlib.py", line 170, in wrapper
    result = func(*args, **kwargs)
  File ".../glue/glue/viewers/matplotlib/state.py", line 24, in notify
    super(DeferredDrawCallbackProperty, self).notify(*args, **kwargs)
  File ".../echo/core.py", line 125, in notify
    cback(new)
  File ".../jdaviz/jdaviz/configs/imviz/plugins/viewers.py", line 182, in on_limits_change
    self.set_compass(self.state.layers[i].layer)
  File ".../jdaviz/jdaviz/configs/imviz/plugins/viewers.py", line 188, in set_compass
    self.compass = self.session.parent_app.get_tray_item_from_name('imviz-compass')
AttributeError: 'Session' object has no attribute 'parent_app'
```